src/main/haskell/kore/test/Test/Kore/Step/Substitution.hs

```diff
@@ -0,0 +1,213 @@
+                [   ( Mock.x
+                    , Mock.constr10 Mock.a
+                    )
+                ]
+            )
+        )
+
+    , testCase "Constructor normalization with variables"
+        -- [x=constructor(y)] + [x=constructor(y)]  === [x=constructor(y)]
+        (assertEqual ""
+            ( Right
+                ( PredicateSubstitution
+                    makeTruePredicate
+                    [   ( Mock.x
+                        , Mock.constr10 (mkVar Mock.y)
+                        )
+                    ]
+                )
+            )
+            ( normalize
+                [   ( Mock.x
+                    , Mock.constr10 (mkVar Mock.y)
+                    )
+                ]
+                [   ( Mock.x
+                    , Mock.constr10 (mkVar Mock.y)
+                    )
+                ]
+            )
+        )
+
+    , testCase "Double constructor is bottom"
+        -- [x=constructor(a)] + [x=constructor(constructor(a))]  === bottom?
+        -- TODO(Vladimir) the result is error instead of bottom
+        (assertEqual ""
+            ( Left
+                ( UnificationError
+                    ( PatternClash
+                        ( HeadClash Mock.aSymbol )
+                        ( HeadClash Mock.constr10Symbol )
+                    )
+                )
+            )
+            ( normalize
+                [   ( Mock.x
+                    , Mock.constr10 Mock.a
+                    )
+                ]
+                [   ( Mock.x
+                    , Mock.constr10 (Mock.constr10 Mock.a)
+                    )
+                ]
+            )
+        )
+
+    , testCase "Double constructor is bottom with variables"
+        -- [x=constructor(y)] + [x=constructor(constructor(y))]  === bottom?
+        -- TODO(Vladimir) the result is error instead of bottom, and different
+        --   than the one without variables.
+        (assertEqual ""
+            ( Left (UnificationError NonFunctionalPattern) )
+            ( normalize
+                [   ( Mock.x
+                    , Mock.constr10 (mkVar Mock.y)
+                    )
+                ]
+                [   ( Mock.x
+                    , Mock.constr10 (Mock.constr10 (mkVar Mock.y))
+                    )
+                ]
+            )
+        )
+
+    , testCase "Constructor and constructor of function errors"
+        -- [x=constructor(a)] + [x=constructor(f(a))]  === error
+        (assertEqual ""
+            ( Left (UnificationError (NonConstructorHead Mock.fSymbol)) )
+            ( normalize
+                [   ( Mock.x
+                    , Mock.constr10 Mock.a
+                    )
+                ]
+                [   ( Mock.x
+                    , Mock.constr10 (Mock.f Mock.a)
+                    )
+                ]
+            )
+        )
+
+    , testCase "Constructor and constructor of function errors with variables"
+        -- [x=constructor(y)] + [x=constructor(f(y))]  === error
```

Comment:
Could you also add a test with a functional symbol instead of f? Also, is the TODO below needed?